runtime/onert/core/src/ir/train/UseDefGenerator.cc

```diff
@@ -0,0 +1,187 @@
+#include "../verifier/Verifier.h"
+
+#include <cassert>
+#include <memory>
+
+// TODO Reduce duplicate code
+
+namespace onert
+{
+namespace ir
+{
+namespace train
+{
+
+UseDefGenerator::UseDefGenerator(const TrainableGraph &tgraph)
+  : _tgraph{tgraph}, _node_to_idx{}, _training_usedefs{}
+{
+  const auto order = _tgraph.topolSortOperations();
+  for (const auto &index : order)
+  {
+    const auto &node = _tgraph.operation(index);
+    assert(_node_to_idx.find(&node) == _node_to_idx.end());
+    _node_to_idx[&node] = index;
+  }
+
+  // Check whether loss exists
+  assert(std::any_of(order.begin(), order.end(),
+                     [&](const auto &index) {
+                       return _tgraph.operation(index).opcode() == ir::OpCode::Loss;
+                     }) &&
+         "Loss does not exist");
+}
+
+UseDefChains UseDefGenerator::operator()()
+{
+  const auto &graph = _tgraph.graph();
+  assert(ir::verifier::EdgeChecker().verify(graph));
+
+  _training_usedefs.clear();
+  graph.operands().iterate([&](const ir::OperandIndex &idx, const ir::Operand &operand) {
+    // Initialize as emtpy UseDefChain
+    const auto empty_usedef_chain = UseDefChain{operand};
+    _training_usedefs.emplace(TrainingOperandIndex{idx, true}, empty_usedef_chain);
+    _training_usedefs.emplace(TrainingOperandIndex{idx, false}, empty_usedef_chain);
+  });
+
+  initForForwardingNodes();
+
+  initForBackwardingNodes();
+
+  return _training_usedefs;
+}
+
+void UseDefGenerator::visit(const train::operation::Loss &node)
+{
+  assert(_node_to_idx.find(&node) != _node_to_idx.end());
+  const auto &op_index = _node_to_idx.at(&node);
+  const auto backwarding_op_index = TrainingOperationIndex{op_index, false};
+
+  for (const auto &in_index : node.getInputs() | ir::Remove::UNDEFINED | ir::Remove::DUPLICATED)
+  {
+    // Insert use of forwarding inputs
+    const auto in_forwarding_index = TrainingOperandIndex{in_index, true};
+    insertUse(in_forwarding_index, backwarding_op_index);
+  }
+
+  // Set def of backwarding(backprop) y_pred
+  const auto &y_pred_index = node.getInputs().at(train::operation::Loss::Input::Y_PRED);
+  assert(!_tgraph.operands().at(y_pred_index).isConstant());
+  const auto y_pred_outgoing_index = TrainingOperandIndex{y_pred_index, false};
+  insertBackPropDef(y_pred_outgoing_index, backwarding_op_index);
+
+  // Set def of backwarding(backprop) y_true
+  const auto &y_true_index = node.getInputs().at(train::operation::Loss::Input::Y_TRUE);
+  assert(!_tgraph.operands().at(y_true_index).isConstant());
+  const auto y_true_outgoing_index = TrainingOperandIndex{y_true_index, false};
+  insertBackPropDef(y_true_outgoing_index, backwarding_op_index);
+
+  // Remove use of backwarding output
+  const auto &out_index = node.getOutputs().at(0);
+  const auto incoming_index = TrainingOperandIndex{out_index, false};
+  auto &usedef_chain = _training_usedefs.at(incoming_index);
+  usedef_chain.removeTrainingUse(backwarding_op_index);
```

Comment:
Why it is removed? 

```
Y_pred (in)    ---    LOSS    --- output (out) 
[y_pred_index]       /           [out_index] 
                   / 
Y_true (in)   ---/
[y_true_index] 
```

Is `out_index`(`ouput` of LOSS node) not used while backwarding?

Reply:
Yes it's not used in any loss operations(CategoricalCrossentropy, MeanSquaredError) yet. Those loss layers uses forwarding `y_pred` instead of forwarding `output` in backwarding.

Reply:
Is this code directly related to removing the acual output_index memory?

Reply:
`output` has a def as forwarding loss op. So output memory is always allocated.

Reply:
@ragmani Yes, this memory is used to get the loss value using the `nnfw_train_get_loss` function. So if this memory is removed through this code, this code should be modified.

Reply:
Even so, currently, there is no problem in my draft including this code because those codes do not release or reuse that memory.
I was curious about it because I consider it if I change planning each non-const tensors, back-prop tensors to plan together.